zebra-chain/src/orchard/tree.rs

```diff
@@ -154,70 +153,104 @@ impl ZcashDeserialize for Root {
-                height: 0,
-                count: 0,
-            };
+impl NoteCommitmentTree {
+    /// Adds a note commitment x-coordinate to the tree.
+    ///
+    /// The leaves of the tree are actually a base field element, the
+    /// x-coordinate of the commitment, the data that is actually stored on the
+    /// chain and input into the proof.
+    ///
+    /// Returns an error if the tree is full.
+    #[allow(clippy::identity_op)]
+    pub fn append(&mut self, cm_x: pallas::Base) -> Result<(), ()> {
+        if self.is_complete() {
+            // Tree is full
+            return Err(());
         }
 
-        let count = values.len() as u32;
-        let mut height = 0u8;
-        let mut current_layer: VecDeque<pallas::Base> = values.into_iter().collect();
-
-        while usize::from(height) < MERKLE_DEPTH {
-            let mut next_layer_up = vec![];
-
-            while !current_layer.is_empty() {
-                let left = current_layer.pop_front().unwrap();
-                let right;
-                if current_layer.is_empty() {
-                    right = EMPTY_ROOTS[height as usize];
-                } else {
-                    right = current_layer.pop_front().unwrap();
+        match (self.left, self.right) {
+            (None, _) => self.left = Some(cm_x),
+            (_, None) => self.right = Some(cm_x),
+            (Some(l), Some(r)) => {
+                let mut combined =
+                    merkle_crh_orchard((MERKLE_DEPTH - 1 - 0) as u8, Some(l), Some(r));
+                self.left = Some(cm_x);
+                self.right = None;
+
+                for i in 0..MERKLE_DEPTH {
+                    if i < self.parents.len() {
+                        if let Some(p) = self.parents[i] {
+                            combined = merkle_crh_orchard(
+                                (MERKLE_DEPTH - 1 - (i + 1)) as u8,
+                                Some(p),
+                                combined,
+                            );
+                            self.parents[i] = None;
+                        } else {
+                            self.parents[i] = combined;
+                            break;
+                        }
+                    } else {
+                        self.parents.push(combined);
+                        break;
+                    }
                 }
-                next_layer_up.push(merkle_crh_orchard(height, Some(left), Some(right)).unwrap());
             }
-
-            height += 1;
-            current_layer = next_layer_up.into();
         }
 
-        assert!(current_layer.len() == 1);
+        Ok(())
+    }
 
-        NoteCommitmentTree {
-            root: Root(current_layer.pop_front().unwrap()),
-            height,
-            count,
-        }
+    /// Returns the current root of the tree, used as an anchor in Orchard
+    /// shielded transactions.
+    #[allow(clippy::identity_op)]
+    pub fn root(&self) -> Root {
```

Comment:
This code was copied from librustzcash - we want to avoid modifying it as much as possible.

```suggestion
    /// shielded transactions.
    //
    // This code is based on https://github.com/zcash/librustzcash/blob/4c4d226f404aef64df6a77caa289b30cc875ae4f/zcash_primitives/src/merkle_tree.rs#L160
    #[allow(clippy::identity_op)]
    pub fn root(&self) -> Root {
```